### PR TITLE
Mainactivity track packagename change

### PR DESCRIPTION
**Pull request by wedgberto**
#

### Platforms affected
Android

### Motivation and Context

Cordova has built in hook functionality that enables customisation of the build - my project has used the hook feature to alter the package name via command line arguments so that side by side versions of the app can be built. e.g. io.cordova.helloword can be installed and ran alongside io.cordova.helloword2 because they have different package names. This is very useful for building a test, UAT or beta release. This was possible in cordova-android@8.1.0 but was broken by changes made in 9.0.0. When the package name of the project is changed in config.xml, the MainActivity.Java file is no longer moved to a folder that tracks the package name and ends up being removed.

This leads to an app crash when launched because the MainActivity class does not exist.

Subsequent build attempts of the project fail because the MainActivity.java file is missing from the project.

#1139 

### Description

This change reverts the behaviour of ensuring that MainActivity.java is located in a folder that tracks the package name read from config.xml. I.e a package name of io.cordova.helloworld in config.xml results in MainActivity.java existing in src/main/java/io/cordova/helloworld even if the package name in config.xml is altered after the platform is added to the project.

### Testing

I have added a new Jasmine unit test "relocate MainActivity.java", but need some help proving that this test is correct.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [ ] I've updated the documentation if necessary

### CODE REVIEW

File: bin/templates/cordova/lib/prepare.js [hunk 1]
```diff
@@ -249,14 +249,18 @@ function updateProjectAccordingTo (platformConfig, locations) {
...
 
-    const destFile = java_files[0];
-
-    // var destFile = path.join(locations.root, 'app', 'src', 'main', 'java', androidPkgName.replace(/\./g, '/'), path.basename(java_files[0]));
-    // fs.ensureDirSync(path.dirname(destFile));
-    // events.emit('verbose', java_files[0]);
-    // events.emit('verbose', destFile);
-    // console.log(locations);
-    // fs.copySync(java_files[0], destFile);
+    var destFile = java_files[0];
```
Comment (breautek) [inline bin/templates/cordova/lib/prepare.js:252]: For new code we are trying to migrate to using `let` or `const` over `var` for variable declarations. So here we should use `let` instead.

```suggestion
    let destFile = java_files[0];
```

File: bin/templates/cordova/lib/prepare.js [hunk 2]
```diff
@@ -249,14 +249,18 @@ function updateProjectAccordingTo (platformConfig, locations) {
...
+
+    // if package name has changed, path to MainActivity.java has to track it
+    var newDestFile = path.join(locations.root, 'app', 'src', 'main', 'java', androidPkgName.replace(/\./g, '/'), path.basename(java_files[0]));
```
Comment (breautek) [inline bin/templates/cordova/lib/prepare.js:255]:
```suggestion
    let newDestFile = path.join(locations.root, 'app', 'src', 'main', 'java', androidPkgName.replace(/\./g, '/'), path.basename(java_files[0]));
```

File: bin/templates/cordova/lib/prepare.js [hunk 3]
```diff
@@ -249,14 +249,18 @@ function updateProjectAccordingTo (platformConfig, locations) {
...
+    if (newDestFile !== destFile) {
+        fs.ensureDirSync(path.dirname(newDestFile));
+        events.emit('verbose', destFile);
+        events.emit('verbose', newDestFile);
+        console.log(locations);
```
Comment (breautek) [inline bin/templates/cordova/lib/prepare.js:260]: Console usage should not remain in the codebase.

```suggestion
```

File: spec/unit/prepare.spec.js [hunk 4]
```diff
@@ -81,9 +82,9 @@ function mockGetIconItem (data) {
     }, data);
 }
 
-describe('prepare', () => {
-    describe('updateIcons method', function () {
-    // Rewire
+describe('prepare', function () {
+    describe('updateIcons', function () {
+        // Rewire
```
Comment (breautek) [inline spec/unit/prepare.spec.js:87]: Arrow functions shouldn't converted to `function ()` style.

Additionally this change appears to be unrelated to the problem this PR is trying to solve. PRs should be concise to only include changes required to address a problem.

File: spec/unit/prepare.spec.js [hunk 5]
```diff
@@ -493,7 +494,7 @@ describe('prepare', () => {
         });
     });
 
-    describe('prepareIcons method', function () {
+    describe('prepareIcons', function () {
```
Comment (breautek) [inline spec/unit/prepare.spec.js:497]: Arrow functions shouldn't converted to `function ()` style.

Additionally this change appears to be unrelated to the problem this PR is trying to solve. PRs should be concise to only include changes required to address a problem.

File: spec/unit/prepare.spec.js [hunk 6]
```diff
@@ -833,4 +834,78 @@ describe('prepare', () => {
             ).toBeResolved();
         });
     });
+
+    describe('updateProjectAccordingTo', function () {
+    describe('relocate MainActivity.java', function () {
```
Comment (breautek) [inline spec/unit/prepare.spec.js:839]: sub describe blocks should have an indentation level. Functions should be in an arrow function format.

```suggestion
    describe('updateProjectAccordingTo', () => {
    describe('relocate MainActivity.java', () => {
```

File: test/android/app/build.gradle [hunk 7]
```diff
@@ -25,7 +25,7 @@ android {
     buildToolsVersion "29.0.2"
 
     defaultConfig {
-        applicationId "org.apache.cordova.unittests"
+        applicationId "org.apache.cordova.unittests2"
```
Comment (breautek) [inline test/android/app/build.gradle:28]: Why is this change required?

File: test/android/app/src/main/java/org/apache/cordova/unittests/StandardActivity.java [hunk 8]
```diff
@@ -17,7 +17,7 @@ Licensed to the Apache Software Foundation (ASF) under one
        under the License.
 */
 
-package org.apache.cordova.unittests;
+package org.apache.cordova.unittests2;
```
Comment (breautek) [inline test/android/app/src/main/java/org/apache/cordova/unittests/StandardActivity.java:20]: Why is this change required?

File: bin/templates/gradle.properties [hunk 9]
```diff
@@ -0,0 +1,4 @@
+org.gradle.daemon=true
+org.gradle.jvmargs=-Xmx2048m
+android.useAndroidX=false
+android.enableJetifier=false
```
Comment (breautek) [inline bin/templates/gradle.properties:4]: Why is this file being added here?

File: spec/unit/prepare.spec.js [hunk 10]
```diff
@@ -833,4 +834,78 @@ describe('prepare', () => {
...
+
+        // Spies
+        let replaceFileContents;
+
+        // Mock Data
+        let cordovaProject;
+        let options;
+        let packageName;
+
+        beforeEach(function () {
```
Comment (breautek) [inline spec/unit/prepare.spec.js:853]:
```suggestion
        beforeEach(() => {
```

**Review by breautek — CHANGES_REQUESTED** [review]
Thank you for your time and effort into investigating and preparing this PR. Unfortunately this is a PR that cannot be accepted as it currently stands.

I've noted a number of changes that I suggest. Generally speaking, we should try to avoid changing code that is not relevant to the problem that the PR is trying to resolve.

Additionally, there appears to be a syntax error in `prepare.spec.js` that is preventing eslint and other tests to completely run.

```
> eslint . "bin/**/!(*.*|gitignore)"


D:\a\cordova-android\cordova-android\spec\unit\prepare.spec.js
  912:1  error  Parsing error: Unexpected token

✖ 1 problem (1 error, 0 warnings)

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! cordova-android@9.1.0-dev lint: `eslint . "bin/**/!(*.*|gitignore)"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the cordova-android@9.1.0-dev lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\npm\cache\_logs\2020-12-03T15_46_34_987Z-debug.log
npm ERR! Test failed.  See above for more details.
Error: Process completed with exit code 1.
```

Please look into this. You can run the tests locally by running `npm test`.

Kind regards,
Norman

File: bin/templates/cordova/lib/prepare.js (hunk 3 above)
Comment (wedgberto) [reply bin/templates/cordova/lib/prepare.js:260]: I was trying to figure out what was going on with the unit test

File: spec/unit/prepare.spec.js (hunk 6 above)
Comment (wedgberto) [reply spec/unit/prepare.spec.js:839]: This was a consequence of basing this test on an existing test

File: test/android/app/src/main/java/org/apache/cordova/unittests/StandardActivity.java (hunk 8 above)
Comment (wedgberto) [reply test/android/app/src/main/java/org/apache/cordova/unittests/StandardActivity.java:20]: this PR is about being able to change the package name. I must have changed this for that reason but I don't remember if it is in fact needed so I will revert it

File: bin/templates/gradle.properties (hunk 9 above)
Comment (wedgberto) [reply bin/templates/gradle.properties:4]: this is an unintended artefact of the new unit test. I will remove it

File: spec/unit/prepare.spec.js (hunk 10 above)
Comment (wedgberto) [reply spec/unit/prepare.spec.js:853]: another copy/paste artefact

File: bin/templates/gradle.properties (hunk 9 above)
Comment (wedgberto) [reply bin/templates/gradle.properties:4]: the new unit test generates this file every time it runs

File: test/android/app/build.gradle (hunk 7 above)
Comment (wedgberto) [reply test/android/app/build.gradle:28]: reverted

File: spec/unit/prepare.spec.js (hunk 5 above)
Comment (wedgberto) [reply spec/unit/prepare.spec.js:497]: all non related changes have been reverted

File: spec/unit/prepare.spec.js (hunk 4 above)
Comment (wedgberto) [reply spec/unit/prepare.spec.js:87]: all non related changes have been reverted

**Comment by wedgberto** [comment]
@breautek Thanks for the gentle comments. I have implemented your suggestions, but would like to request that this PR not be accepted until I have the new unit test working as intended.

**Comment by erisu** [comment]
Is there a reason why the branch of this PR is based off the 9.x branch/tag?

You have release related commits in this PR that is not in master.

File: bin/templates/gradle.properties (hunk 9 above)
Comment (breautek) [reply bin/templates/gradle.properties]: You may be able to use spies to prevent actual code logic from running. I fear if unit tests causes side effects such as creating files, then they may not run in a consistent manner if the file already exists.

**Comment by wedgberto** [comment]
I am going to abandon this PR and start again

**Comment by wedgberto** [comment]
this PR has been abandoned and a new PR created

**Comment by wedgberto** [comment]
> Is there a reason why the branch of this PR is based off the 9.x branch/tag?
> 
> You have release related commits in this PR that is not in master.

I did that by mistake. I've created a new PR that was branched from master - #1154. 
This issue is breaking the automated build of my app so I would really appreciate the new PR being reviewed.

**Comment by wedgberto** [comment]
@bre

> @breautek Thanks for the gentle comments. I have implemented your suggestions, but would like to request that this PR not be accepted until I have the new unit test working as intended.

@breautek I abandoned this PR and started a new one #1154  ('cos I base this this one from the wrong branch as highlighted by @erisu)
